src/sentry/models/file.py

```diff
@@ -274,17 +309,38 @@ def seek(self, pos):
     def tell(self):
         if self.closed:
             raise ValueError('I/O operation on closed file')
+        if self.prefetched:
+            return self._curfile.tell()
+        if self._curfile is None:
+            return self.size
         return self._curidx.offset + self._curfile.tell()
 
-    def read(self, bytes=4096):
+    def read(self, n=-1):
         if self.closed:
             raise ValueError('I/O operation on closed file')
-        result = ''
-        while bytes and self._curfile is not None:
-            blob_result = self._curfile.read(bytes)
-            if not blob_result:
-                self._nextidx()
-                continue
-            bytes -= len(blob_result)
-            result += blob_result
-        return result
+
+        if self.prefetched:
+            return self._curfile.read(n)
```

Comment:
I'm doubtful that this works correct, so please help me understand here. :)

Right now, when I say "prefetch this file", we start by kicking off 4 concurrent blobs at a time and they all get written into a file concurrently as their chunks stream in.

But nothing here seems to wait for any of the chunk futures to actually be completed. So let's imagine:

I have a file who has 100 chunks (100MB file). We blast in to preload all of the chunks. We immediately seek to chunk 50, then read. We'd get no data, or the seek would fail or something in between, not entirely sure of the API.

But I feel like we need some other wrapper around the temporary file to keep track of all of the futures to only read the chunks as they come in.

Or even if simply chunk 1 takes 500ms to load, but we try to read it immediately. It'll be empty in the tempfile.

I feel like in practice this is going to be more problematic without anything actually waiting on the chunks to be finished and handing back the temporary file's handler as-is.

Reply:
I think i clarified this on chat. The executor joins.